[PT-D] Reshard API for ShardedTensor and PartialTensor

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69735
* __->__ #70079

We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

Commit:
[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor

We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -539,3 +542,11 @@ def wrapper(*args, **kwargs):
+        module: torch.nn.Module,
+        resharding_spec: ReshardingSpec) -> ModuleResharder:
+    """
+    Replaces a module with a new module that will reshard its output.
```
For now, probably make this a private API `_reshard_output` since probably the public API will be at the Sharder level. Also, we need additional docs here in terms of `Args:` and `Returns:`.

nit: "Replaces a module with a new module that will reshard its output according to the given ``resharding_spec``".

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
 
+    def reshard(self, resharding_spec: ReshardingSpec) -> Union[ShardedTensor, torch.Tensor]:
+        """
+        Reshards a tensor given the resharding_spec, if output_local_tensors is
```
nit: Use double ticks `` to highlight `output_local_tensors`

fduwjj replied:
Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
+        True in resharding_spec.
+
+        It would output the local shard on each rank instead of ShardedTensor
+        (only works for a single local shard).
```
nit: `Args:` and `Returns:`

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
+        # TODO: implement the reshard for ShardedTensor to ShardedTensor.
+        output_local_tensors = resharding_spec.output_local_tensors
+        if not output_local_tensors:
+            raise RuntimeError(
```
nit: use `NotImplementedError`

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
+            raise RuntimeError(
+                "Reshard not supported for ShardedTensor output yet."
+            )
+        from torch.distributed._sharded_tensor.ops._common import (
+            _result_distribute_with_col_rearrange,
+        )
+        local_shard = self.local_shards()[0].tensor
+        local_results = []
+        for i in range(local_shard.size(0)):
+            local_results.append(local_shard[i].t().contiguous())
+
+        # Distribute results to each rank with col rearrangement.
+        output = _result_distribute_with_col_rearrange(
+            local_results,
+            local_shard,
+            len(self._metadata.shards_metadata),
+            self,
+            self._process_group,
+        )
+
+        # transpose the output and return result.
+        return output.t().contiguous()
```
We are not using the `resharding_spec` here at all to determine how to reshard?

fduwjj replied:
For aggregation to a local tensor... Probably yes..

pritamdamania87 replied:
> For aggregation to a local tensor... Probably yes..

Not sure I followed your comment. Don't we need to use `resharding_spec` here to make this function generic? To begin with, I think supporting a handful of resharding cases which fit our use cases is fine and we throw errors in other cases. But we need to have a generic implementation here.

fduwjj replied:
I see.. Sure, will do. Thanks!

fduwjj replied:
Make the reshard a generic function.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
     def __torch_function__(self, func, types, args=(), kwargs=None):
         if func in _SHARDED_OPS:
             return _SHARDED_OPS[func](types, args, kwargs, self._process_group)
+        elif func == torch.nn.functional.gelu:
+            kwargs['op'] = torch.nn.functional.gelu
+            return elementwise_ops(types, args, kwargs)
```
Why aren't we just using `sharded_op_impl` decorator here similar to other ops?

fduwjj replied:
Let me rebase on top of your landed PR.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+
+class PartialTensor(object):
+    """
+    PartialTensor is an abstraction to represent Tensors that need
+    aggregation across multiple devices and multiple processes.
```
We should mention that the PartialTensor `shard` on each rank is of the same size and give a few examples of aggregation functions like `add` to combine all shards of a PartialTensor.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    for all the shards.
+
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard
+    and the global metadata.
```
There is no global metadata in the current implementation?

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+
+
+class PartialTensor(object):
```
make this class private for now since it is mostly for our own internal use.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+            Partial results are stored across ranks and in a form of a list.
+    """
+
+    def __init__(self, local_partial_results):
```
We need an additional enum here for type of aggregation needed.

fduwjj replied:
sure, let me update the code to reflect the design.

fduwjj replied:
I think we can put that enum into resharding spec. LKM if this sounds make sense to you or not.

pritamdamania87 replied:
Hmm I don't think the type of aggregation fits in the resharding spec. ReshardingSpec probably only mentions how to reshard the data but shouldn't care about how to aggregate the data into a single Tensor.

fduwjj replied:
I see. I think what you mean here because the aggregation is kind of exclusive to the partial tensor. Another idea I have is that, can we directly use the enum `ReduceOp` instead of creating a new one ourselves?

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    torch.ones, or torch.empty.
+
+    Args:
+        local_partial_results (List(Tensor)):
```
What are the semantics here? If we aggregate do we get a List[Tensor] of aggregated results by aggregating each index? I think it is simpler to just allow only a single Tensor per rank.

fduwjj replied:
So the local result is a list of tensors. After aggregation, each rank will have one tensor because we will use `reduce_scatter`. Using one tensor also works, we then need to stack or concat them into a single tensor. We can then specify the dim in the resharding spec. Let me change it.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+            if not isinstance(tensor, torch.Tensor):
+                raise ValueError('Element in local_partial_results needs to be a Tensor.')
+
+    def reshard(self, _resharding_spec: ReshardingSpec) -> torch.Tensor:
```
nit: why does `_resharding_spec` have an underscore? (usually this is done for hidden params)

fduwjj replied:
Oh, with the enum added, we will be using it.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        Consolidate all partial tensors across multiple ranks based on resharding
+        spec.
+        """
+        local_result = torch.empty_like(self.local_partial_results[0])
+        return reduce_scatter(local_result, self.local_partial_results)
```
This only works if aggregation type is `ADD`.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+
+    def reshard(self, _resharding_spec: ReshardingSpec) -> torch.Tensor:
+        """
+        Consolidate all partial tensors across multiple ranks based on resharding
+        spec.
```
nit: `Args:` and `Returns:` needed.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -750,3 +796,61 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        return reduce_scatter(local_result, self.local_partial_results)
+
+
+class ModuleResharder(torch.nn.Module):
```
make this private too since it is for internal use.

pritamdamania87 replied:
Ah this is indeed a much better solution since it keeps the the original module intact as well. @fduwjj We should try out this hook to see if it fits our use case.

fduwjj replied:
Sure. Let me look into it.

fduwjj replied:
Can we use this hook for the prod use? The wiki says "This adds global state to the nn.module module and it is only intended for debugging/profiling purposes.". We probably need to go with our previous proposal..

fduwjj replied:
Done. The hook way works.

pritamdamania87 commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -41,6 +41,34 @@ class ShardingSpec(PlacementSpec):
+@dataclass
+class ReshardingSpec(object):
+    """
+    This is a wrapper of PlacementSpec that defines the new placement as being
```
There is no PlacementSpec here, it is a wrapper of ShardingSpec.

fduwjj replied:
Sure. I first thought we have a PlacementSpec inside the ShardingSpec. Let me change it.

pritamdamania87 commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -41,6 +41,34 @@ class ShardingSpec(PlacementSpec):
+    This is a wrapper of PlacementSpec that defines the new placement as being
+    resharded across multiple devices or aggregate into a local tensor.
+
+    The semantics of how a tensor is partitioned is inline with
+    :meth:`torch.chunk`, where ``dim`` in torch.chunk corresponds to the
+    specified ``dim`` and ``chunks`` in torch.chunk is the number of elements
+    in the placement specified.
```
This is not true, ReshardingSpec has a ShardingSpec which could be anything. This comment is referring to `ChunkShardingSpec` which might not always be the case.

fduwjj replied:
sure, let me change it.

pritamdamania87 commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -41,6 +41,34 @@ class ShardingSpec(PlacementSpec):
+            The spec contains the new dimension to reshard on and placement.
+        output_local_tensors(bool):
+            Whether or not to output the local shards of the ShardedTensor as
+            a tensor.
```
nit: ``torch.Tensor``, use double ticks (``) around ShardedTensor and torch.Tensor to emphasize.

pritamdamania87 commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -41,6 +41,34 @@ class ShardingSpec(PlacementSpec):
+    output_local_tensors: bool = False
+
+    def __post_init__(self):
+        if not isinstance(self.spec, ShardingSpec):
```
Probably also check `ShardingSpec` is instance of `ChunkShardingSpec` since we will only support that for now.

pritamdamania87 reviewed (commented):
Thanks for writing up this proposal! I think it would be great to actually use these APIs in our unit test and demonstrate how this would look from a user point of view.

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
 
+    def reshard(self, resharding_spec: ReshardingSpec) -> Union[ShardedTensor, torch.Tensor]:
+        """
+        Reshards a tensor given the resharding_spec, if output_local_tensors is
```
shouldn't this be "reshard to a tensor"? Also, maybe we want a better explanation on what does it mean of "reshard to a tensor", does it mean we only output local shards, or does it mean we all_gather the local shards and make that as an output, seems a bit confusing

fduwjj replied:
Sure, let me make the comment more clear. We return the results after `all_gather`.

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -550,9 +559,46 @@ def sharding_spec(self) -> ShardingSpec:
         """
         return self._sharding_spec
 
+    def reshard(self, resharding_spec: ReshardingSpec) -> Union[ShardedTensor, torch.Tensor]:
```
@fduwjj @pritamdamania87 I feel like "reshard" seems a bit confusing to user. it seems to me that "output_local_tensors" case we will convert a sharded tensor to a data parallel tensor, is this a "reshard" behavior? or is it a "to_replicated_tensor" or "all_gather_shards"? I feel like we aren't "reshard" the tensor from a user prospective.

pritamdamania87 replied:
@wanchaol There are two parts here, the first part is `resharding` which is done as per the provided ShardingSpec. Now the second part is `output_local_tensors`, this essentially means that we want to return the local tensors after this operation. For example, this is very useful in cases where the rest of the code is DataParallel and works only with Tensors and not ShardedTensors. Given that ShardedTensors have a very limited set of ops, resharding and converting to Tensors would be very useful in certain cases.

wanchaol replied:
Yeah "resharding" API is necessary I think, what I mentioned is the second part of it, basically `output_local_tensors` is kinda fused with the resharding, but I feel like it's a different behavior and we should separate them tbh. The second part is essentially a conversion from ShardedTensor to Data Parallel tensors, it's not a "resharding" behavior, as the output is not another ShardedTensor, rather, it's a ReplicatedTensor. I can't think of a rational API of this yet, maybe sth like `ShardedTensor.to_replicated()`? I just feel like a fused behavior between these two is a bit confusing.

wanchaol replied:
Also if we separate these two behaviors, I think we could get rid of the "ReshardingSpec" and just use ShardingSpec universally, which might be a bit cleaner.

fduwjj replied:
For this use case, it is indeed not a resharding strictly speaking. It's not trying to converted to a replicate, neither. But resharding of sharded tensor is still needed as @pritamdamania87 has pointed out. So what if we enable the resharding for Partial Tensor for now and create a `to_dp_tensor` method for sharded tensor? What do you guys think? @wanchaol @pritamdamania87

fduwjj replied:
After thinking this for a while, I think this indeed is a cleaner option. We don't need to define a new resharding spec. For the API's naming, I guess this is more like a reverse action of sharding op. (aka. the reverse op of api `sharded_param`.) Maybe let's call it `collect_local_shards` (open for a better name, too) ? We then can use this API for partial tensor. 

So basically, for Partial Tensor or sharded Tensor -> Tensor, we can directly call this API to do the convert. If we want to covert one sharded Tensor to a new sharded tensor, we then need `reshard_output`.

Does this sound clearer and less confusing? @wanchaol and @pritamdamania87

wanchaol replied:
Yeah that make sense to me, either `extract_output_tensors` or `collect_local_shards` works for me, agreed we will need to implement build a higher level api on them in the Sharder API. and yeah we don't need ReshardingSpec, could use ShardingSpec universally.

pritamdamania87 replied:
`collect_local_shard` sounds good to me. I think it is simpler to have the API deal with just one local shard rather than return `List[Tensor]`, because in most use cases we will actually have just a single Tensor and if we return List[Tensor], there needs to be additional logic somewhere to convert it to a single Tensor for most existing data parallel code.

pritamdamania87 replied:
cc @mrshenli If you have any better ideas/suggestions around the API :)

wanchaol reviewed (commented):
overall looks good, thanks for implementing this! Some question/suggestions on the API inlined. Also, can we add a test to test this API, or do you think the test will be almost the same as the PR stacked above?

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

fduwjj commented:
1. Update the design.
2. Add all unit tests. (The two wrapper API will be covered in the next PR's unit test.)
3. Address all comments.

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+            tensor([[3], [3], [5], [5], [7], [7], [9], [9]]) # Rank 2
+            tensor([[4], [4], [6], [6], [8], [8], [10], [10]]) # Rank 3
+        """
+        if (
```
Shouldn't we validate single local shard here?

fduwjj replied:
Oh, I see. Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        tensor([3, 4]) # Rank 0
+        tensor([3, 4]) # Rank 1
+
+        >>> # All tensors below are of torch.cfloat type.
+        >>> # We have 2 process groups, 2 ranks.
+        >>> tensor = torch.tensor([1+1j, 2+2j], dtype=torch.cfloat) + 2 * rank * (1+1j)
+        >>> tensor = torch.stack([tensor, tensor])
+        >>> tensor
+        tensor([[1.+1.j, 2.+2.j], [1.+1.j, 2.+2.j]]) # Rank 0
+        tensor([[3.+3.j, 4.+4.j], [3.+3.j, 4.+4.j]]) # Rank 1
+        >>> partial_tensor = _PartialTensor(tensor)
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([4.+4.j, 6.+6.j]) # Rank 0
+        tensor([4.+4.j, 6.+6.j]) # Rank 1
```
Any reason to add complex tensor examples here?

fduwjj replied:
Oh, I copied from the reduce_scatter comment. If this is not necessary, I can remove it. But this should work for complex cases, I can add unit test cases for that.

fduwjj replied:
Update: turns out that complex number is not supported. I will add a value Error for now.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    """
+
+    def __init__(self, local_partial_shard, pg, aggr_op=distributed_c10d.ReduceOp.SUM):
+        self.local_partial_shard = local_partial_shard
```
nit: local_shard is probably sufficient

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        tensor([4.+4.j, 6.+6.j]) # Rank 1
+    """
+
+    def __init__(self, local_partial_shard, pg, aggr_op=distributed_c10d.ReduceOp.SUM):
```
make pg optional `pg=None` and initialize self.pg to the default group if pg is None.

fduwjj replied:
Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    def reshard(self, resharding_spec: ShardingSpec) -> ShardedTensor:
+        """
+        Consolidate all partial tensors across multiple ranks and covert to a sharded
+        tensor based on the resharding spec.
```
Lets add some details here in terms of logically what the result of the operation is:

1. We first aggregate all the shards of the partial tensor.
2. We then shard this tensor according to the provided spec.

Also, add a note that these are logical steps but for performance reasons actual implementation might combine two steps together.

fduwjj replied:
Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        """
+        sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
+        world_size = dist.get_world_size(self.pg)
+        local_shards = torch.tensor_split(self.local_partial_shard, world_size)
```
Any reason not to use `torch.chunk` here?

fduwjj replied:
I just don't know this api. Let me change it.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
+        world_size = dist.get_world_size(self.pg)
+        local_shards = torch.tensor_split(self.local_partial_shard, world_size)
+        local_result = reduce_scatter(
+            torch.empty_like(local_shards[0]), list(local_shards), op=self.aggr_op
+        )
```
This only works if sharding_dim is 0 right? Don't we need to take into account `resharding_spec.dim` here before we perform appropriate aggregation?

fduwjj replied:
I guess Partial Tensor does not have the info of the original sharding dim. Or do we assume that we chunk by the `resharding_dim`?

fduwjj replied:
Update: After a second thought and reading through the logical comment. I think we can use `resharding_spec.dim` for chunk.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+
+        rank = dist.get_rank(self.pg)
+        local_metadata = None
+        shard_tensor_size = list(local_result.size())
```
nit: sharded_tensor_size

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        rank = dist.get_rank(self.pg)
+        local_metadata = None
+        shard_tensor_size = list(local_result.size())
+        shard_tensor_size[sharding_dim] *= world_size
```
Can't we just do `sharded_tensor_size = local_partial_shard.size()`?

fduwjj replied:
Oh, right. Let me change it.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+                )
+                shards.append(Shard(local_result, local_metadata))
+                break
+            current_offsets[sharding_dim] += local_result.size(sharding_dim)  # type: ignore[index]
```
Does this work for uneven sharding across ranks? For example if this code is run on the last rank where `local_result.size()` is probably smaller than the rest on `sharding_dim`?

fduwjj replied:
I thought about these cases for a while. For now, we have two assumptions:
1. World size needs to divide the length of the dimension to be resharded.
2. Each tensor has the same size across all ranks. 

The reason for these assumptions is that Partial Result comes from row-wise sharding linear op. It's not a sharded tensor and we don't contain a sharding spec here for how it is sharded previously. Plus, it looks like `reduce_scatter` does not support uneven size (different ranks have different tensor sizes). So I guess for now we are good?

pritamdamania87 replied:
Can't we just use `get_chunked_dim_size` and `get_split_size` to get the right size for each partition given uneven split? We know the size of the complete tensor and which dim it is being sharded on so it should be possible to figure out the size/offset on each rank based on that instead of assuming world_size needs to divide length of dimension.

> Plus, it looks like reduce_scatter does not support uneven size (different ranks have different tensor sizes).

I'm assuming this only applies to the same index position in the list of inputs. But different indices in the input list can have different sizes?

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -329,6 +330,187 @@ def test_shard_parameter_errors(self):
+    @with_comms(init_rpc=False)
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_partial_tensor_reshard(self):
```
Lets put PartialTensor tests in a separate file.

fduwjj replied:
Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -538,3 +539,40 @@ def wrapper(*args, **kwargs):
 # Import all builtin sharded ops
 from .ops import *  # noqa: F403
+
+def _reshard_output(
+        module: torch.nn.Module,
+        resharding_spec: ShardingSpec) -> torch.nn.Module:
+    """
+    Hook a module with local shards collection in the forward pass according
+    to the given ``resharding_spec``.
+
+    Args:
+        module (:class:`torch.nn.Module`): Module whose output needs to be resharded.
+        resharding_spec (:class:`torch.distributed._sharding_spec.ShardingSpec`): The
+            specification describing how the sharded tensor will be resharded.
+
+    Returns:
+        A :class:`torch.nn.Module` object with collection API hooked.
+    """
+    def hook_func(_module, _input, output):
+        return output.reshard(resharding_spec)
+    module.register_forward_hook(hook_func)
+    return module
+
+
+def _collect_local_shard(module: torch.nn.Module) -> torch.nn.Module:
```
We need tests for these two new APIs too.

fduwjj replied:
Do we feel the test in the sharded_linear and megatron prototype is enough?

fduwjj replied:
Update: After a second thought, I think it still makes sense to create a dedicated unit test for them. I will update the unit test plan.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -329,6 +330,187 @@ def test_shard_parameter_errors(self):
+            dim=0,
+            placements=ranks,
+        )
+        ranks[0], ranks[1], ranks[2], ranks[3] = ranks[3], ranks[0], ranks[1], ranks[2]
```
Can we just shuffle the list using random.shuffle instead?

fduwjj replied:
I have tried that but the tricky thing is that this is called across ranks. And each rank will shuffle it to a different order. So I have to do the shuffle manually.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -329,6 +330,187 @@ def test_shard_parameter_errors(self):
+
+class TestReshard(ShardedTensorTestBase):
+    def _run_sharded_tensor_reshard(self, sharding_spec, reshard_spec, input_size):
+        torch.manual_seed(0)
+        fc = torch.nn.Linear(*input_size).cuda(self.rank)
+        fc_compare = torch.nn.Linear(*input_size).cuda(self.rank)
+        fc_compare.weight = torch.nn.Parameter(fc.weight.detach().clone())
+        shard_parameter(fc, "weight", sharding_spec)
+        st = fc.weight
+        st.reshard(reshard_spec)
+        shard_parameter(fc_compare, "weight", reshard_spec)
+        st_compare = fc_compare.weight
```
Do we need all this if we are just testing ShardedTensor reshard? We can just create a ShardedTensor directly instead of using nn.Linear and `shard_parameter`?

fduwjj replied:
Sure. I was taking a short-cut here. Let me change it.

fduwjj replied:
Update: Will create a new API for tensor sharding.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -329,6 +330,187 @@ def test_shard_parameter_errors(self):
+        shard_parameter(fc_compare, "weight", reshard_spec)
+        st_compare = fc_compare.weight
+        self.assertEqual(1, len(st.local_shards()))
+        self.assertEqual(
+            st.local_shards()[0].tensor,
+            st_compare.local_shards()[0].tensor,
+        )
```
Can we compare the metadata too?

fduwjj replied:
Sure.

fduwjj replied:
Update: Turns out I did not have the logic of updating the metadata. Just added one and test cases in unit test.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -329,6 +330,187 @@ def test_shard_parameter_errors(self):
+    @requires_nccl()
+    def test_partial_tensor_reshard(self):
+        spec = ChunkShardingSpec(
+            dim=0,
```
We should test with dim=1 as well

fduwjj replied:
Oh, I changed the dim below.

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -329,6 +330,187 @@ def test_shard_parameter_errors(self):
+            tensor = torch.rand(*input_size).cuda(self.rank)
+            results.append(tensor)
+            results_compare.append(tensor.clone().detach())
+        pg = dist.new_group(ranks=ranks)
```
Why do we need a new_group? We can just use the default group since we are anyways using all ranks?

fduwjj replied:
I see. Sure.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -746,3 +923,120 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        Returns:
+            A :class:`ShardedTensor` filled with local aggregated result.
+        """
+        sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
```
We should validate that `resharding_spec` is `ChunkShardingSpec`.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+        self.local_shards()[0].tensor = local_shard
+        return self
+
+    def collect_local_shard(self) -> torch.Tensor:
```
Sure. Just a quick question here, is it high-pri to support multiple local shard on one rank? Or maybe that use case is not very common?

pritamdamania87 replied:
I think for sharded tensor ops we should simplify and only support one local shard for now.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+            not isinstance(self._sharding_spec, ChunkShardingSpec)
+        ):
+            raise NotImplementedError("Only ChunkShardingSpec supported for reshard.")
+        current_sharding_dim = int(self._sharding_spec.dim)  # type: ignore[attr-defined]
+        reshard_dim = int(resharding_spec.dim)
```
Yes, but mypy complains about it so I added the conversion.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+                for idx, placement in enumerate(self._sharding_spec.placements):  # type: ignore[attr-defined]
+                    if current_rank == placement.rank():  # type: ignore[union-attr]
+                        new_dim = resharding_spec.placements[idx].placement.rank()  # type: ignore[union-attr]
+                        input_split_sizes[new_dim] = local_shard.size(0)
```
oh, this is kind of tricky. Because for this case, we don't any split and send the whole local shard to another rank. Like a point-to-point communication because the sharding dim is same as resharding dim. Only placements are different.

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

fduwjj commented:
Addressed comments from reviewers and split the `shard_parameter` API into two: `shard_tensor` and `shard_parameter` so that we can shard a local tensor based on the spec and have the code logic reused by `shard_parameter`.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+                gathered_input_size = list(local_shard.size())
+                gathered_input_size[reshard_dim] = sharded_dim_size
+                gathered_input = torch.empty(gathered_input_size, device=local_shard.device)
+                local_shard = all_to_all_single(
```
This comment is not addressed?

fduwjj replied:
Opps, Github collapses 10+ comments which I thought were already addressed.. Let me change the rest of them.

fduwjj replied:
Update: this is indeed needed. Looks like the autograd enabled API does not change the tensor in place and we need to update it via the returned tensor.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+                    input_split_sizes=input_split_sizes,
+                    group=self._process_group,
+                )
+                return self
```
Doesn't look like this was addressed either?

fduwjj replied:
Good catch. I also realized that I missed that part.

fduwjj replied:
Update: Turns out the initial test cases do not fully cover this logic. I will update the correct one.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -370,25 +370,19 @@ def pre_load_state_dict_hook(module, state_dict, prefix, local_metadata, strict,
-    :class:`torch.distributed._sharded_tensor.ShardedTensor`
-
     Args:
         module (:class:`torch.nn.Module`): Module whose parameter needs to be sharded.
```
This is incorrect, it should be torch.Tensor.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -400,23 +394,17 @@ def shard_parameter(
             the default process group will be used.
 
+    Returns:
+        A :class:`ShardedTensor` shared from the given tensor.
```
nit: shared -> sharded

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -400,23 +394,17 @@ def shard_parameter(
+        A :class:`ShardedTensor` shared from the given tensor.
+
     .. warning::
         Only :class:`torch.distributed._sharding_spec.ShardingSpec` is
```
This should be ChunkShardingSpec, we should fix the docs in `shard_parameter` too

fduwjj replied:
I see. So we directly assume the input is a `ChunkShardingSpec`? Don't we need to make it more generic?

fduwjj replied:
Update: Opps. Misreading the comment and will correct the error in both places.

pritamdamania87 reviewed (commented):
Changes look good overall, although I'm wondering if some local commits were not pushed? A bunch of my previous comments were marked resolved but I don't see corresponding code changes (especially in _sharded_tensor/api.py).

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+        local_shard = self.local_shards()[0].tensor
+        current_rank = dist.get_rank(self._process_group)
+        world_size = dist.get_world_size(self._process_group)
+        if current_sharding_dim == reshard_dim:
```
I think moving to a separate file is a better idea. Let me do it.

fduwjj replied:
Update: Still keep the main body here to avoid dependency loop and move the most of logic to the file `reshard.py`.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+                return self
+
+        # Compute expected size
+        split_size = get_split_size(local_shard.size(reshard_dim), world_size)
+        input_split_sizes = [0] * world_size
+        rearrange_input = False
+        for idx, placement in enumerate(resharding_spec.placements):  # type: ignore[attr-defined]
+            sharded_dim_size = get_chunked_dim_size(
+                local_shard.size(reshard_dim), split_size, idx
+            )
+            input_split_sizes[placement.rank()] = sharded_dim_size  # type: ignore[union-attr, index]
+            if placement.rank() != idx:  # type: ignore[union-attr]
+                rearrange_input = True
+
+        if rearrange_input:
+            # Need to re-arrange reshard_dim of local_shard before all2all.
+            indices: List[List[int]] = [[0]] * world_size
+            # When we do the chunk split, we always ensure the first N - 1 chunks get max out
+            # and then the Nth chunk gets the rest. So input_split_sizes like [3, 3, 3, 4]
+            # are not possible. The expected split size will be [4, 4, 4, 1].
+            sharded_dim_size_max = max(input_split_sizes)
+            for idx, placement in enumerate(resharding_spec.placements):
+                split_size = input_split_sizes[placement.rank()]  # type: ignore[union-attr, index]
+                offset_start_idx = idx * sharded_dim_size_max
+                indices[placement.rank()] = list(  # type: ignore[union-attr, index]
+                    range(offset_start_idx, offset_start_idx + split_size)
+                )
+            indices_flatten = list(idx for indice in indices for idx in indice)
+            local_shard = local_shard.index_select(
+                reshard_dim, torch.tensor(indices_flatten, device=local_shard.device)
+            )
```
Let me see if I can consolidate some part of it. There are indeed subtle differences so we might not be able to consolidate them all.

fduwjj replied:
Update: Move the major logic into `reshard.py` and split it into two major functions. For further refactoring, I can send a separate PR to address that.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -546,6 +553,176 @@ def sharding_spec(self) -> ShardingSpec:
+
+        # Because reshard_dim != original shard_dim. We need to compute the
+        # size of tensor from each rank.
+        output_tensor_list = [torch.tensor(1)] * world_size
```
We can do that. Just because of mypy warnings. I guess I can use another ignore here.

mrshenli commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    PartialTensor doesn't provide any Tensor like operations but is a
+    wrapper providing the Tensor representing the local partial shard.
+
+    We assume the size of each local tensor to be exactly the same.
```
If the world size does not divide the length of the dimension, would I be correct that we will do padding? Or do we rely on users to make sure this assumption is always met?

fduwjj replied:
This is a good point. The initial design only considers the case when we store the result of row-wise sharding linear op in each rank. Let me add a check here in the reshard and make the comment clearer.

mrshenli commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+
+    Args:
+        local_partial_shard (Tensor): Partial result stored across ranks.
+        pg (ProcessGroup): The process group to aggregate on.
```
nit: naming abbreviation style a bit inconsistent. Besides, DDP called this arg `process_group`.

fduwjj replied:
Got it. Will change it to `process_group`.

mrshenli commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+    Args:
+        local_partial_shard (Tensor): Partial result stored across ranks.
+        pg (ProcessGroup): The process group to aggregate on.
+        aggr_op (distributed_c10d.ReduceOp): Way to aggregate the partial result.
```
nit: `all_reduce` API call this `op`. If you prefer to be more precise, `reduce_op` might be better, as it would match the `ReduceOp` type name.

fduwjj replied:
OK, thanks for your suggestion. Let me change it.

mrshenli commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+            )
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([3, 4]) # Rank 0
+        tensor([3, 4]) # Rank 1
```
If the inputs are

```python
tensor([[1, 2], [3, 4]]) # Rank 0
tensor([[5, 6], [7, 8]]) # Rank 1
```

would I be correct that the output will then be:

```python
tensor([5, 6]) # Rank 0
tensor([7, 8]) # Rank 1
```

If so, shall we update the example. It might help make it more explicit.

mrshenli replied:
curious, in real-world applications, when are we going to use min/max/prod ops? (I think I can understand the SUM use case :) )

fduwjj replied:
If inputs are:
`tensor([[1, 2], [3, 4]]) # Rank 0
tensor([[5, 6], [7, 8]]) # Rank 1`

The outputs will be:
`tensor([6, 8]) # Rank 0
tensor([10, 12]) # Rank 1`.

I will update the comment to make it clearer.

fduwjj replied:
One thing I can think of is the max/min pooling might leverage this?

mrshenli replied:
> The outputs will be:
> tensor([6, 8]) # Rank 0 tensor([10, 12]) # Rank 1.

I might have missed sth. I saw the example was using `ReduceOp.MAX` instead of `ReduceOp.SUM`. In that case, will it still do a sum?

fduwjj replied:
Oh, yes. In the case of `MAX`, your understanding is correct. I will update the comment.

mrshenli commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+            torch.empty_like(local_shards[0]), list(local_shards), op=self.aggr_op
+        )
+
+        rank = dist.get_rank(self.pg)
```
`self.pg.rank` should also work:

https://github.com/pytorch/pytorch/blob/b36b11cbc1f11fa805894927ca667443f3618a34/torch/csrc/distributed/c10d/init.cpp#L964-L965

fduwjj replied:
Got it. Thanks!!

mrshenli commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+            )
+        >>> complete_tensor = partial_tensor.reshard(collect_spec)
+        >>> complete_tensor
+        tensor([3, 4]) # Rank 0
```
The `reshard()` API says it returns a `ShardedTensor`. But this print reads like a normal local tensor. 

I'd assume the output should be sth like `ShardedTensor...`?

https://github.com/pytorch/pytorch/blob/b36b11cbc1f11fa805894927ca667443f3618a34/torch/distributed/_sharded_tensor/api.py#L649-L650

fduwjj replied:
Yeah the output should be a sharded tenor. This is local tensor. Let me update the comment to reflect that.

wanchaol commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        st = ShardedTensor._init_from_local_shards(
+            shards, tuple(shard_tensor_size), process_group=self.pg
+        )
+        st._sharding_spec = copy.deepcopy(resharding_spec)
```
nit: why do we need to deepcopy the resharding_spec?

fduwjj replied:
Just on the safe side the pass-in spec might be changed later on. Are you concerned about the extra cost from deepcopy?

wanchaol replied:
yeah but I guess the cost is pretty small and it indeed avoid the potential of changing the pass-in spec.

wanchaol reviewed (commented):
lgtm, some nits and suggestions inline.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -551,6 +558,176 @@ def sharding_spec(self) -> ShardingSpec:
+        self.local_shards()[0].tensor = local_shard
+        return self
+
+    def local_shard(self) -> torch.Tensor:
```
Or maybe name it `local_shard_tensor`? We can avoid multiple calls like `local_shards()[0]`. cc: @pritamdamania87

wanchaol replied:
yeah it's a common operation, `local_shard_tensor` lgtm.

pritamdamania87 replied:
The purpose was to avoid a lot of duplicate code we have today where we validate there is a single shard and then do `local_shards()[0].tensor`. I think instead of `local_shard_tensor`, maybe we just call it `local_tensor`.

fduwjj replied:
I see. OK, let's use a shorter name then.

fduwjj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -538,3 +579,40 @@ def wrapper(*args, **kwargs):
+        A :class:`torch.nn.Module` object with collection API hooked.
+    """
+    def hook_func(_module, _input, output):
+        return output.reshard(resharding_spec)
```
Sure.

fduwjj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -369,25 +370,19 @@ def pre_load_state_dict_hook(module, state_dict, prefix, local_metadata, strict,
-def shard_parameter(
-        module: torch.nn.Module,
-        param_name: str,
+def shard_tensor(
```
I think this can be a public API because the use case is kind of common? How do we decide whether to make it public or private? Out of curiosity and for my learning purpose.

fduwjj replied:
Update: after a second thought, I will make this a private API for now.

wanchaol replied:
I was thinking that this might cause confusion to the user on which API to use. But yeah this is indeed a common API, maybe we can make it public later if we find it's necessary

fduwjj commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -487,6 +476,58 @@ def shard_parameter(
+        currently supported as the ``sharding_spec``.
+    """
+    # Perform some validation first.
+    if not isinstance(sharding_spec, ChunkShardingSpec):
```
Sure.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -551,6 +558,176 @@ def sharding_spec(self) -> ShardingSpec:
+                        "rank:3/cuda:3",
+                    ],
+                )
+            >>> st = _sharded_tensor.rand(spec, 8, 2)
+            >>> st.local_shards()[0].tensor = tensor
```
You are right. `_init_from_local_shards ` is a better fit. Will change it.

fduwjj commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -751,3 +928,122 @@ def _create_tensor_from_params(*size, local_device, tensor_init_params: TensorIn
+        shard_tensor_size[sharding_dim] *= world_size
+        current_offsets = [0] * len(local_result.size())
+        shards = []
+        for idx, placement in enumerate(resharding_spec.placements):  # type: ignore[attr-defined]
```
For now, yes. `resharding_spec.placements` follows the ranks. Let me update the code.

pritamdamania87 replied:
@fduwjj This is not necessary though right? placements could be [rank3, rank2, rank0, rank1]?

fduwjj replied:
@pritamdamania87  Right, I take a look at the unit test of sharded linear op, we did shuffle the ranks in the placement. We cannot make the assumption here.

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D][RFC] Resharding related API implement for ShardedTensor and Partial Tensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

fduwjj commented:
Addressed comments from reviewers:
1. Split reshard code and reshard/parital tensor test into different files.
2. Fix logic errors in reshard and make the comments in partial tensor clearer.
3. Ensure every comment has been resolved or discussed.
4. Will rebase on https://github.com/pytorch/pytorch/pull/71742 once it is closed.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -596,7 +591,9 @@ def _reshard_output(
     """
     def hook_func(_module, _input, output):
-        return output.reshard(resharding_spec)
+        if isinstance(output, ShardedTensor) or isinstance(output, _PartialTensor):
```
Output can probably be a tuple too? I guess we should handle that case and reshard all elements of the tuple that are ShardedTensor or PartialTensor. We should also update the docs to mention that we reshard only ShardedTensor and PartialTensor.

fduwjj replied:
Oh, I see. I have not thought about the case of tuple. Maybe we need to check if it's an Iterable?

fduwjj replied:
Filed an issue here to follow up: https://github.com/pytorch/pytorch/issues/72022.

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -1018,18 +985,20 @@ def reshard(self, resharding_spec: ShardingSpec) -> ShardedTensor:
         sharding_dim = int(resharding_spec.dim)  # type: ignore[attr-defined]
-        world_size = dist.get_world_size(self.pg)
-        local_shards = torch.tensor_split(self.local_partial_shard, world_size)
+        if self.local_shard.size(sharding_dim) % self.process_group.size() != 0:
+            raise ValueError('World size need to divide the length of the dimension.')
```
Why do we need this restriction?

fduwjj replied:
This is to answer @mrshenli 's previous comment. I guess this is one assumption we have for PartialTensor right? Otherwise, we will get uneven shape after `torch.chunk`?

fduwjj replied:
Discussed offline, we are going to add padding for these cases and I included in the filed GH issue: https://github.com/pytorch/pytorch/issues/72022

pritamdamania87 reviewed (approved):
There are a bunch of minor items we need to address (ex: multiple outputs of a module, PartialTensor reshard where sharding dim is not divisible by world_size etc.). But lets create gh issues for all follow up items and address them in PRs on top of this.

fduwjj commented:
Follow-up records in this GH issue: https://github.com/pytorch/pytorch/issues/72022.

Commit:
Update on "[PT-D] Reshard API for ShardedTensor and PartialTensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D] Reshard API for ShardedTensor and PartialTensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

Commit:
Update on "[PT-D] Reshard API for ShardedTensor and PartialTensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]

fduwjj commented:
Rebased on top of https://github.com/pytorch/pytorch/pull/72141#pullrequestreview-870034365.

Commit:
Update on "[PT-D] Reshard API for ShardedTensor and PartialTensor"


We defined a new concept named `PartialTensor`, which is an abstraction to represent Tensors that need aggregation across multiple devices and multiple processes.

We also defined a API `reshard_output` to reshard a `PartialTensor` to `Tensor` or reshard a `ShardedTensor` to `ShardedTensor/Tensor`. This is done via class `ModuleResharder` which acts like a wrapper of original modules plus the a reshard in the final step.

The `reshard` logic is defined in each class (`ShardedTensor` and `PartialTensor`).

Differential Revision: [D33121037](https://our.internmc.facebook.com/intern/diff/D33121037/)

[ghstack-poisoned]
